### DISCUSSION TITLE
Samsung apparently included StrongSWan on their Android Build

### DISCUSSION THREAD

**Discussion by aryasenna** [strongswan/strongswan#1439] [General]
Maybe this is a common knowledge? I was trying to troubleshoot IKEv2 connection on the Android native client, and  was surprised instead of Racoon or[ AOSP IPSec module](https://android.googlesource.com/platform/packages/modules/IPsec/) , I got Charon on my ADB logs!

```
11-24 16:53:16.853 12088  5472 I charon  : 09[DMN] setting up TUN device for CHILD_SA android{164}
11-24 16:53:16.932 12088  5472 I charon  : 09[DMN] successfully created TUN device
11-24 16:53:16.932 12088  5472 I charon  : 09[IKE] peer supports MOBIKE
11-24 16:53:18.086 12088 12355 I charon  : 00[IKE] deleting IKE_SA android[180] between 192.168.128.21[user]...xx.xx.xx.xx[xxx.net]
11-24 16:53:18.086 12088 12355 I charon  : 00[IKE] sending DELETE for IKE_SA android[180]
11-24 16:53:18.086 12088 12355 I charon  : 00[ENC] generating INFORMATIONAL request 6 [ D ]
11-24 16:53:18.088 12088 12355 I charon  : 00[NET] sending packet: from 192.168.1xx.xx[57667] to xx.xx.xx.xx[4500] (65 bytes)
11-24 16:53:21.593 12088 12355 I charon  : 00[LIB] loaded plugins: androidbridge charon android-log openssl fips-prf random nonce pubkey chapoly curve25519 pkcs1 pkcs8 pem xcbc hmac socket-default revocation eap-identity eap-mschapv2 eap-md5 eap-gtc eap-tls x509`
```

Anyone know what version  this is? I suspect this is somewhere between  5.8.1 and  5.9.1

Since bug [ #3557](https://wiki.strongswan.org/issues/3557) still persists: effectively all public-CA fails to validate since no public OCSP server returns a nonce,

**Comment by tobiasbrunner** [comment]
Hm, that log looks more like that of our [Android app](https://docs.strongswan.org/docs/5.9/os/androidVpnClient.html) (it has the _androidbridge_ component loaded but not the _kernel-netlink_ plugin and it creates a TUN device). Is this actually from the built-in client in the Samsung phone? (It's known that the latter actually uses strongSwan, but it's surprising that it does this in this way.)

Anyway, if the built-in client actually has that bug, you have to contact Samsung so they can fix it in their build.

**Reply by aryasenna** [reply]
Yup positive, I don't have the StrongSwan app installed!

It kinda surprises not even the maintainer know exactly how Strongswan code is shipped with probably millions of  Samsung devices.😀 IANAL, but Samsung should consider commercial license, no? 

> Anyway, if the built-in client actually has that bug, you have to contact Samsung so they can fix it in their build.

I can try, they have bug report applet, but probably this is something to be fixed in major update (and they just have released Android 13..)

**Reply by tobiasbrunner** [reply]
> It kinda surprises not even the maintainer know exactly how Strongswan code is shipped with probably millions of Samsung devices. IANAL, but Samsung should consider commercial license, no?


It is known and they do have a commercial license. What surprises me is that they apparently use it like an app with TUN devices and userland IPsec (they didn't do that previously). But if that works for them, fine :) Might make it even easier to update as it is possibly not integrated in the system image but some userland component.

A workaround is obviously to use our own app.

**Reply by aryasenna** [reply]
Actually you are right, I was also testing strongswan app after seeing that log. I must have got the logs mixed up. 

The bixby routine (Samsung built-in Tasker) does not support 3rd party VPN.. also I imagine native client is easier on the battery, and I don't have to deal with Doze vs Wakelock to make MOBIKE/DPD/KeepAlive works reliably

This is the _actual_ charon from built-in client.

```
00[KNL] USE_KERNEL_NETLINK_IPSEC
00[LIB] created TUN device: ipsec
00[LIB] loaded plugins: charon android-log openssl fips-prf nonce pkcs1 pkcs8 pem xcbc hmac kernel-netlink socket-default eap-identity eap-mschapv2 xauth-generic android kernel-libipsec revocation x509 curl unity constraints
```

**Reply by tobiasbrunner** [reply]
Yeah, that log makes more sense :) (But interesting that they load the _kernel-libipsec_ plugin. It won't get used if the _kernel-netlink_ plugin is loaded first, but they might have the option to use the userland IPsec implementation for some scenarios.)

> The bixby routine (Samsung built-in Tasker) does not support 3rd party VPN.

If you can send custom Intents, you can start/stop connections with our client (see the page I linked above).

> also I imagine native client is easier on the battery, and I don't have to deal with Doze vs Wakelock to make MOBIKE/DPD/KeepAlive works reliably

Possible, but if the app is exempted from the power restrictions (it prompts the user) it should work fine and not have much of an impact unless there is actual traffic (or the connection has to get reestablished etc.).
